File: scala/Continuations.scala
import Spec._
import Undefined._

object Continuations {

  def getUser(req: Request)(cont: User => Response): Response = {
    val token = req.header.get("Authorization")
    if (token.isEmpty) noToken else
    if (`malformed token?`) malformedToken(token.get) else
    if (`user not found?`) noUser(token.get) else
    cont(`the user`)
  }

  def getResource(req: Request)(cont: Resource => Response): Response = {
    val path: String = req.path
    if (`resource not found?`) noResource(path) else
    cont(`the resource`)
  }

  def execute(content: String, usr: User, src: Resource)
             (cont: Unit => Response): Response = {
    if (!`is permitted?`) notPermitted(src.path) else
    if (!`is executed?`) badConnection else
    cont(())
  }

  def handlePost(req: Request): Response = {
    if (req.method != "POST") notAllowed(req.method) else
    if (req.body.isEmpty) noBody else
    getUser(req) { usr =>
    getResource(req) { src =>
    execute(req.body, usr, src) { _ =>
    success(req.path, req.body) } } }
  }
}

File: haskell/Continuations.hs
module Continuations where

import Spec

getUser :: Request -> (User -> IO Response) -> IO Response
getUser (Request _ _ _ header) cont =
  case lookup "Authorization" header of
    Nothing -> return noToken
    Just token ->
      if is_malformed_token then return (malformedToken token) else
      if is_user_not_found then return (noUser token) else
      cont the_user

getResource :: Request -> (Resource -> IO Response) -> IO Response
getResource (Request path _ _ _) cont =
  is_resource_not_found_io >>= \notFound ->
  if notFound then return (noResource path) else
  the_resource_io >>= cont

execute :: String -> User -> Resource -> (() -> IO Response) -> IO Response
execute body usr src@(Resource path) cont =
  is_permitted_io >>= \permitted ->
  if not permitted then return (notPermitted path) else
  is_executed_io >>= \executed ->
  if not executed then return badConnection else
  cont ()

handlePost :: Request -> IO Response
handlePost req@(Request path method body _) =
  if method /= "POST" then return (notAllowed method) else
  if null body then return noBody else
  getUser req (\usr ->
  getResource req (\src ->
  execute body usr src (\_ ->
  return (success path body))))
